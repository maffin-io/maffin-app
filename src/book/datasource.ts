import initSqlJs from 'sql.js';
import { DataSource } from 'typeorm';
import { useSWRConfig } from 'swr';

import {
  Account,
  Book,
  Commodity,
  Price,
  Split,
  Transaction,
} from './entities';

export async function initDB(database: Uint8Array): Promise<DataSource> {
  await initSqlJs({
    // Required to load the wasm binary asynchronously. Of course, you can host it wherever you want
    locateFile: file => `https://sql.js.org/dist/${file}`,
  });

  let datasource: DataSource;

  if (database.length !== 0) {
    datasource = new DataSource({
      type: 'sqljs',
      synchronize: true,
      database,
      logging: false,
      entities: [Account, Book, Commodity, Price, Split, Transaction],
    });
    await datasource.initialize();
  } else {
    datasource = new DataSource({
      type: 'sqljs',
      synchronize: true,
      logging: false,
      entities: [Account, Book, Commodity, Price, Split, Transaction],
    });
    await datasource.initialize();
    await createEmptyBook();
  }

  return datasource;
}

// TODO: Need to mutate the accounts SWR key so Root is shown!
async function createEmptyBook() {
  const rootAccount = Account.create({
    guid: 'rootAccount',
    name: 'Root',
    type: 'ROOT',
  });
  await Account.upsert(rootAccount, ['guid']);

  await Book.upsert(
    {
      guid: 'maffinBook',
      fk_root: rootAccount.guid,
    },
    ['guid'],
  );
}
