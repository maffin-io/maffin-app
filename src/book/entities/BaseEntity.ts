import {
  BaseEntity as BE,
  BeforeInsert,
  BeforeUpdate,
  PrimaryColumn,
  SaveOptions,
} from 'typeorm';
import * as v from 'class-validator';
import { QueryClient } from '@tanstack/react-query';

/**
 * Override BaseEntity so all entities have a generated guid in the format
 * we want and adds validation before saving or updating entities
 */
export default class BaseEntity extends BE {
  static CACHE_KEY = '';

  @BeforeInsert()
  @BeforeUpdate()
  async validate() {
    const errors = await v.validate(this, { stopAtFirstError: true });
    if (errors.length) {
      throw new Error(`validation failed: ${errors}`);
    }
  }

  @PrimaryColumn({
    type: 'varchar',
    length: 32,
  })
  @v.IsString()
    guid: string = crypto.randomUUID().substring(0, 31);

  async save(options?: SaveOptions): Promise<this> {
    const entity = await super.save(options);

    if (this.queryClient && ['Account', 'Commodity'].includes(this.constructor.name)) {
      updateCache({ queryClient: this.queryClient, entity });
    }

    return entity;
  }

  async remove(options?: SaveOptions): Promise<this> {
    const entity = await super.remove(options);

    if (this.queryClient && ['Account', 'Commodity'].includes(this.constructor.name)) {
      updateCache({ queryClient: this.queryClient, entity, isDelete: true });
    }

    return entity;
  }

  get queryClient(): QueryClient | undefined {
    // @ts-ignore
    return this.constructor.dataSource.options.extra?.queryClient;
  }
}

/**
 * Updates `/api/<entity_name>` and
 * `/api/<entity_name>/<guid>` keys so changes are propagated
 * in the cache.
 *
 * - /api/<entity_name> is updated as a list of Entity[]
 * - /api/<entity_name>/<guid> is updated with the value of the Entity
 */
export async function updateCache(
  {
    queryClient,
    entity,
    isDelete = false,
  }: {
    queryClient: QueryClient,
    entity: BaseEntity,
    isDelete?: boolean,
  },
) {
  // @ts-ignore
  const key = entity.constructor.CACHE_KEY;

  queryClient.setQueryData(
    [key],
    (entities: BaseEntity[] | undefined) => {
      if (!entities) {
        return undefined;
      }

      const newEntities = [...entities];

      const index = entities.findIndex(e => e.guid === entity.guid);
      if (index === -1) { // New entity added
        newEntities.push(entity);
      } else if (!isDelete) { // Entity updated
        newEntities[index] = entity;
      } else if (isDelete) { // Entity deleted
        newEntities.splice(index, 1);
      }

      return newEntities;
    },
  );

  queryClient.setQueryData(
    [key, { guid: entity.guid }],
    !isDelete ? entity : null,
  );
}
