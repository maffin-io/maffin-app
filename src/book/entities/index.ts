export { default as Account } from './Account';
export { default as BankConfig } from './BankConfig';
export { default as Commodity } from './Commodity';
export { default as Price } from './Price';
export { default as Split } from './Split';
export { default as Transaction } from './Transaction';
export { default as Book } from './Book';
export { default as BaseEntity } from './BaseEntity';
