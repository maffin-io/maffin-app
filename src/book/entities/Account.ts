import * as v from 'class-validator';
import {
  Column,
  Entity, JoinColumn,
  ManyToOne,
  OneToMany,
  RelationId,
  Tree,
  TreeParent,
  TreeChildren,
  SaveOptions,
} from 'typeorm';
import type { QueryClient } from '@tanstack/react-query';

import {
  getAllowedSubAccounts,
  ASSET_ACCOUNTS,
  LIABILITY_ACCOUNTS,
} from '@/book/helpers/accountType';
import type Commodity from './Commodity';
import Split from './Split';
import BaseEntity from './BaseEntity';

/**
 * CREATE TABLE accounts (
 *   guid            CHAR(32) PRIMARY KEY NOT NULL,
 *   name            text(2048) NOT NULL,
 *   account_type    text(2048) NOT NULL,
 *   commodity_guid  CHAR(32) NOT NULL,
 *   commodity_scu   integer NOT NULL,
 *   non_std_scu     integer NOT NULL,
 *   parent_guid     CHAR(32),
 *   code            text(2048),
 *   description     text(2048),
 *   hidden          integer NOT NULL,
 *   placeholder     integer NOT NULL
 * );
 */

@Entity('accounts')
@Tree('nested-set')
export default class Account extends BaseEntity {
  static CACHE_KEY = ['api', 'accounts'];
  static TYPES = [
    'ROOT',
    'EQUITY',
    'INCOME',
    'EXPENSE',
    ...ASSET_ACCOUNTS,
    ...LIABILITY_ACCOUNTS,
  ];

  @Column({
    type: 'text',
    length: 2048,
  })
  @v.Matches(/[a-zA-Z.]+/)
  @v.Length(1, 2048)
    name!: string;

  @Column({
    type: 'text',
    enum: Account.TYPES,
    length: 2048,
    name: 'account_type',
  })
  @CheckAccountType()
  @v.IsIn(Account.TYPES)
    type!: string;

  path!: string;

  @TreeParent()
  @JoinColumn({ name: 'parent_guid' })
  @v.IsNotEmpty({ message: 'parent is required' })
  @v.ValidateIf(o => o.type !== 'ROOT')
  // Seems TreeParent doesn't let us create referencing the guid of the account
  // and needs the whole object...
    parent!: Account;

  @RelationId((account: Account) => account.parent)
    parentId: string;

  @TreeChildren()
    children!: Account[];

  @RelationId((account: Account) => account.children)
    childrenIds: string[];

  @ManyToOne('Commodity', { eager: true, cascade: true })
  @JoinColumn({ name: 'commodity_guid' })
  @CheckIECommodity()
  @v.ValidateIf(o => o.type !== 'ROOT')
  @v.IsNotEmpty({ message: 'commodity is required' })
    fk_commodity!: Commodity | string;

  get commodity(): Commodity {
    return this.fk_commodity as Commodity;
  }

  @OneToMany('Split', (split: Split) => split.fk_account)
    splits!: Split[];

  @Column({
    type: 'text',
    length: 2048,
    nullable: true,
  })
  @v.IsOptional()
  @v.Length(4, 2048)
    description?: string;

  @Column({
    default: false,
  })
    hidden!: boolean;

  @CheckPlaceholder()
  @Column({
    default: false,
  })
    // Placeholders are hierarchical accounts that don't containg
    // transactions, they are just parents of other accounts
    placeholder!: boolean;

  async save(options?: SaveOptions): Promise<this> {
    const account = await super.save(options);

    if (this.queryClient) {
      updateCache({ queryClient: this.queryClient });
    }

    return account;
  }

  async remove(options?: SaveOptions): Promise<this> {
    if (this.queryClient) {
      updateCache({ queryClient: this.queryClient });
    }

    const account = await super.remove(options);
    return account;
  }
}

export async function updateCache({ queryClient }: { queryClient: QueryClient }) {
  queryClient.invalidateQueries({
    queryKey: [...Account.CACHE_KEY],
  });
}

// https://github.com/typeorm/typeorm/issues/4714
Object.defineProperty(Account, 'name', { value: 'Account' });

/**
 * Checks allowed types for parent account given the current account's type
 */
function CheckAccountType(validationOptions?: v.ValidationOptions) {
  return function f(object: Account, propertyName: string) {
    v.registerDecorator({
      name: 'checkAccountType',
      target: object.constructor,
      propertyName,
      options: validationOptions,
      validator: {
        validate(type: string, args: v.ValidationArguments) {
          const account = args.object as Account;
          if (account.parent) {
            return getAllowedSubAccounts(account.parent.type).includes(type);
          }

          return true;
        },

        defaultMessage(args: v.ValidationArguments) {
          const account = args.object as Account;
          const allowedTypes = getAllowedSubAccounts(account.parent.type);

          return `only ${allowedTypes} types can be selected with ${account.parent.type} account as parent`;
        },
      },
    });
  };
}

/**
 * Checks that income/expense commodity is the same as the parent.
 *
 * This is a way to check that we always keep the same currency for income/expense
 * account as we only support them in the main currency.
 */
function CheckIECommodity(validationOptions?: v.ValidationOptions) {
  return function f(object: Account, propertyName: string) {
    v.registerDecorator({
      name: 'checkIECommodity',
      target: object.constructor,
      propertyName,
      options: validationOptions,
      validator: {
        validate(fk_commodity: Commodity, args: v.ValidationArguments) {
          const account = args.object as Account;
          if (
            account.parent
            && account.parent.type !== 'ROOT'
            && ['INCOME', 'EXPENSE'].includes(account.type)
          ) {
            return fk_commodity.guid === (account.parent.fk_commodity as Commodity).guid;
          }

          return true;
        },

        defaultMessage(args: v.ValidationArguments) {
          const account = args.object as Account;
          return `Income and Expense accounts must have ${(account.parent.fk_commodity as Commodity).mnemonic} as their commodity`;
        },
      },
    });
  };
}

/**
 * Checks if an account can be a placeholder. Accounts can be Placeholders
 * if they don't have any transactions associated.
 */
function CheckPlaceholder(validationOptions?: v.ValidationOptions) {
  return function f(object: Account, propertyName: string) {
    v.registerDecorator({
      name: 'checkPlaceholder',
      target: object.constructor,
      propertyName,
      options: validationOptions,
      validator: {
        async validate(_: string, args: v.ValidationArguments) {
          const account = args.object as Account;
          if (account.guid) {
            const [, numSplits] = await Split.findAndCount({
              where: { fk_account: { guid: account.guid } },
            });

            if (numSplits > 0 && account.placeholder) {
              return false;
            }
          }

          return true;
        },

        defaultMessage() {
          return 'Placeholder accounts cannot have transactions';
        },
      },
    });
  };
}
