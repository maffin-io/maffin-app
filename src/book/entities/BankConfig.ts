import {
  Column,
  Entity,
  OneToMany,
} from 'typeorm';
import {
  IsString,
} from 'class-validator';

import BaseEntity from './BaseEntity';
import type Account from './Account';

@Entity('bank_config')
export default class BankConfig extends BaseEntity {
  @Column({
    type: 'text',
    length: 2048,
  })
  @IsString()
    token!: string;

  @OneToMany(
    'Account',
    (account: Account) => account.config,
  )
    accounts!: Account[];
}

// https://github.com/typeorm/typeorm/issues/4714
Object.defineProperty(BankConfig, 'name', { value: 'BankConfig' });
