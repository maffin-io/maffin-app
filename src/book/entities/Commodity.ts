import * as v from 'class-validator';
import {
  Column,
  Entity,
} from 'typeorm';

import BaseEntity from './BaseEntity';

/**
 * https://wiki.gnucash.org/wiki/SQL#Tables
 *
 * CREATE TABLE commodities (
 *   guid            CHAR(32) PRIMARY KEY NOT NULL,
 *   namespace       text(2048) NOT NULL,
 *   mnemonic        text(2048) NOT NULL,
 *   fullname        text(2048),
 *   cusip           text(2048),
 *   fraction        integer NOT NULL,
 *   quote_flag      integer NOT NULL,
 *   quote_source    text(2048),
 *   quote_tz        text(2048)
 * );
 */

@Entity('commodities')
export default class Commodity extends BaseEntity {
  @Column({
    type: 'text',
    length: 2048,
  })
  @v.IsString()
  @v.Length(2, 2048)
    namespace!: string;

  @Column({
    type: 'text',
    length: 2048,
  })
  @v.IsString()
  @v.Length(2, 2048)
    mnemonic!: string;

  @Column({
    type: 'text',
    length: 2048,
    default: '',
  })
  @v.IsString()
  @v.IsOptional()
    fullname?: string;

  // We use this field as a way to identify the quote
  // in the exchange wherever we want to pull the price from
  // i.e. can be an ISIN, NVDA, etc. If this is not set, we fallback
  // to mnemonic
  @Column({
    type: 'text',
    length: 2048,
    nullable: true,
    name: 'cusip',
  })
  @v.IsOptional()
  @v.IsString()
    cusip?: string;

  get stockerId(): string {
    return this.cusip || this.mnemonic;
  }
}

// https://github.com/typeorm/typeorm/issues/4714
Object.defineProperty(Commodity, 'name', { value: 'Commodity' });
