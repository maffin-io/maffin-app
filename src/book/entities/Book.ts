import {
  Entity,
  JoinColumn,
  OneToOne,
} from 'typeorm';

import BaseEntity from './BaseEntity';
import Account from './Account';

/**
 * https://wiki.gnucash.org/wiki/SQL#Books
 * CREATE TABLE books (
 *   guid                CHAR(32) PRIMARY KEY NOT NULL,
 *   root_account_guid   CHAR(32) NOT NULL,
 *   root_template_guid  CHAR(32) NOT NULL
 * );
*/

@Entity('books')
export default class Book extends BaseEntity {
  @OneToOne(() => Account, { eager: true })
  @JoinColumn({ name: 'root_account_guid' })
    fk_root!: Account | string;

  get root(): Account {
    return this.fk_root as Account;
  }
}

// https://github.com/typeorm/typeorm/issues/4714
Object.defineProperty(Book, 'name', { value: 'Book' });
