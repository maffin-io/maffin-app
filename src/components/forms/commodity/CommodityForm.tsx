import React from 'react';
import { useForm, Controller } from 'react-hook-form';
import { classValidatorResolver } from '@hookform/resolvers/class-validator';
import { SingleValue } from 'react-select';

import { Commodity } from '@/book/entities';
import Selector from '@/components/selectors/Selector';

const resolver = classValidatorResolver(Commodity, { validator: { stopAtFirstError: true } });

export type FormValues = {
  namespace: string,
  mnemonic: string,
};

export type CommodityFormProps = {
  onSave: Function,
};

const BASE_CURRENCIES = [
  { label: 'EUR' },
  { label: 'USD' },
  { label: 'SGD' },
  { label: 'GBP' },
];

export default function CommodityForm({ onSave }: CommodityFormProps): JSX.Element {
  const form = useForm<FormValues>({
    mode: 'onChange',
    resolver,
  });

  return (
    <form onSubmit={form.handleSubmit((data) => onSubmit(data, onSave))}>
      <fieldset className="text-sm my-5">
        <Controller
          control={form.control}
          name="mnemonic"
          render={({ field, fieldState }) => (
            <>
              <Selector<{ label: string }>
                id="mnemonicInput"
                labelAttribute="label"
                options={BASE_CURRENCIES}
                onChange={(newValue: SingleValue<{ label: string }> | null) => {
                  field.onChange(newValue?.label);
                }}
                placeholder="Choose your currency"
                isClearable={false}
              />
              <p className="invalid-feedback">{fieldState.error?.message}</p>
            </>
          )}
        />
      </fieldset>

      <input
        {...form.register(
          'namespace',
        )}
        hidden
        value="CURRENCY"
        type="text"
      />

      <div className="flex w-full justify-center">
        <button className="btn btn-primary" type="submit" disabled={!form.formState.isValid}>
          Save
        </button>
      </div>
    </form>
  );
}

async function onSubmit(data: FormValues, onSave: Function) {
  const mainCommodity = await Commodity.create({ ...data }).save();
  await onSave(mainCommodity);
}
