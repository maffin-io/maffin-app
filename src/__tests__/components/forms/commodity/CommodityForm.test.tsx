import React from 'react';
import {
  render,
  screen,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { DataSource } from 'typeorm';

import CommodityForm from '@/components/forms/commodity/CommodityForm';
import { Commodity } from '@/book/entities';

describe('CommodityForm', () => {
  let datasource: DataSource;

  beforeEach(async () => {
    datasource = new DataSource({
      type: 'sqljs',
      dropSchema: true,
      entities: [Commodity],
      synchronize: true,
      logging: false,
    });
    await datasource.initialize();
  });

  afterEach(async () => {
    jest.resetAllMocks();
    await datasource.destroy();
  });

  it('renders as expected with add', async () => {
    const { container } = render(
      <CommodityForm
        action="add"
        onSave={() => {}}
      />,
    );

    screen.getByLabelText('Code');
    screen.getByLabelText('Name');
    screen.getByRole('combobox', { name: 'namespaceInput' });
    expect(container).toMatchSnapshot();
  });

  it('renders as expected with update', async () => {
    const { container } = render(
      <CommodityForm
        action="update"
        onSave={() => {}}
      />,
    );

    screen.getByLabelText('Code');
    screen.getByLabelText('Name');
    screen.getByRole('combobox', { name: 'namespaceInput' });
    expect(container).toMatchSnapshot();
  });

  it('renders as expected with delete', async () => {
    const { container } = render(
      <CommodityForm
        action="delete"
        onSave={() => {}}
      />,
    );

    expect(screen.getByLabelText('Code')).toBeDisabled();
    expect(screen.getByLabelText('Name')).toBeDisabled();
    expect(container).toMatchSnapshot();
  });

  it('adds new Commodity and saves', async () => {
    const user = userEvent.setup();
    const mockSave = jest.fn();

    render(
      <CommodityForm
        action="add"
        onSave={mockSave}
      />,
    );

    await user.type(screen.getByLabelText('Code'), 'EUR');
    await user.type(screen.getByLabelText('Name'), 'Euro');
    await user.click(screen.getByRole('combobox', { name: 'namespaceInput' }));
    await user.click(screen.getByText('CURRENCY'));

    expect(screen.getByText('add')).not.toBeDisabled();
    await user.click(screen.getByText('add'));

    const commodities = await Commodity.find();
    expect(commodities).toEqual([
      {
        guid: expect.any(String),
        mnemonic: 'EUR',
        fullname: 'Euro',
        namespace: 'CURRENCY',
        cusip: null,
      },
    ]);
    expect(mockSave).toBeCalledWith({
      guid: expect.any(String),
      mnemonic: 'EUR',
      fullname: 'Euro',
      namespace: 'CURRENCY',
      cusip: null,
    });
  });
});
