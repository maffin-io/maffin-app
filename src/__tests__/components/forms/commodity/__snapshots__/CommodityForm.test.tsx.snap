// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommodityForm renders as expected with add 1`] = `
<div>
  <form>
    <fieldset
      class="my-5 col-start-8 col-span-2"
    >
      <label
        class="inline-block text-sm mb-2"
        for="namespaceInput"
      >
        Type
      </label>
      <div
        class="selector css-b62m3t-container"
        id="namespaceInput"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-2-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
          role="log"
        />
        <div
          class="selector__control css-13cymwt-control"
        >
          <svg
            class="absolute text-md left-[10px]"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 18a7.952 7.952 0 0 0 4.897-1.688l4.396 4.396 1.414-1.414-4.396-4.396A7.952 7.952 0 0 0 18 10c0-4.411-3.589-8-8-8s-8 3.589-8 8 3.589 8 8 8zm0-14c3.309 0 6 2.691 6 6s-2.691 6-6 6-6-2.691-6-6 2.691-6 6-6z"
            />
          </svg>
          <div
            class="selector__value-container css-hlgwow"
          >
            <div
              class="selector__placeholder css-1jqq78o-placeholder"
              id="react-select-2-placeholder"
            >
              &lt;commodity type&gt;
            </div>
            <div
              class="!text-inherit selector__input-container css-19bb58m"
              data-value=""
            >
              <input
                aria-activedescendant=""
                aria-autocomplete="list"
                aria-describedby="react-select-2-placeholder"
                aria-expanded="false"
                aria-haspopup="true"
                aria-label="namespaceInput"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class="selector__input"
                id="react-select-2-input"
                role="combobox"
                spellcheck="false"
                style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="selector__indicators css-1wy0on6"
          >
            <span
              class="hidden selector__indicator-separator css-1u9des2-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class="selector__indicator selector__dropdown-indicator css-1xc3v61-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
        <input
          name="namespaceInput"
          type="hidden"
          value=""
        />
      </div>
      <p
        class="invalid-feedback"
      />
    </fieldset>
    <div
      class="grid grid-cols-12 text-sm gap-2"
    >
      <fieldset
        class="col-span-6"
      >
        <label
          class="inline-block text-sm my-2"
          for="mnemonicInput"
        >
          Code
        </label>
        <input
          class="w-full m-0"
          id="mnemonicInput"
          name="mnemonic"
          type="text"
        />
        <p
          class="invalid-feedback"
        />
      </fieldset>
      <fieldset
        class="col-span-6"
      >
        <label
          class="inline-block text-sm my-2"
          for="fullNameInput"
        >
          Name
        </label>
        <input
          class="w-full m-0"
          id="fullNameInput"
          name="fullname"
          type="text"
        />
        <p
          class="invalid-feedback"
        />
      </fieldset>
    </div>
    <div
      class="flex w-full justify-center mt-5"
    >
      <button
        class="btn capitalize btn-primary"
        type="submit"
      >
        add
      </button>
    </div>
  </form>
</div>
`;

exports[`CommodityForm renders as expected with delete 1`] = `
<div>
  <form>
    <fieldset
      class="my-5 col-start-8 col-span-2"
    >
      <label
        class="inline-block text-sm mb-2"
        for="namespaceInput"
      >
        Type
      </label>
      <div
        class="selector selector--is-disabled css-3iigni-container"
        id="namespaceInput"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-4-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
          role="log"
        />
        <div
          aria-disabled="true"
          class="selector__control selector__control--is-disabled css-16xfy0z-control"
        >
          <svg
            class="absolute text-md left-[10px]"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 18a7.952 7.952 0 0 0 4.897-1.688l4.396 4.396 1.414-1.414-4.396-4.396A7.952 7.952 0 0 0 18 10c0-4.411-3.589-8-8-8s-8 3.589-8 8 3.589 8 8 8zm0-14c3.309 0 6 2.691 6 6s-2.691 6-6 6-6-2.691-6-6 2.691-6 6-6z"
            />
          </svg>
          <div
            class="selector__value-container css-hlgwow"
          >
            <div
              class="selector__placeholder css-1jqq78o-placeholder"
              id="react-select-4-placeholder"
            >
              &lt;commodity type&gt;
            </div>
            <div
              class="!text-inherit selector__input-container css-1mkvw8y"
              data-value=""
            >
              <input
                aria-activedescendant=""
                aria-autocomplete="list"
                aria-describedby="react-select-4-placeholder"
                aria-expanded="false"
                aria-haspopup="true"
                aria-label="namespaceInput"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class="selector__input"
                disabled=""
                id="react-select-4-input"
                role="combobox"
                spellcheck="false"
                style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="selector__indicators css-1wy0on6"
          >
            <span
              class="hidden selector__indicator-separator css-894a34-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class="selector__indicator selector__dropdown-indicator css-1xc3v61-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <p
        class="invalid-feedback"
      />
    </fieldset>
    <div
      class="grid grid-cols-12 text-sm gap-2"
    >
      <fieldset
        class="col-span-6"
      >
        <label
          class="inline-block text-sm my-2"
          for="mnemonicInput"
        >
          Code
        </label>
        <input
          class="w-full m-0"
          disabled=""
          id="mnemonicInput"
          name="mnemonic"
          type="text"
        />
        <p
          class="invalid-feedback"
        />
      </fieldset>
      <fieldset
        class="col-span-6"
      >
        <label
          class="inline-block text-sm my-2"
          for="fullNameInput"
        >
          Name
        </label>
        <input
          class="w-full m-0"
          disabled=""
          id="fullNameInput"
          name="fullname"
          type="text"
        />
        <p
          class="invalid-feedback"
        />
      </fieldset>
    </div>
    <div
      class="flex w-full justify-center mt-5"
    >
      <button
        class="btn capitalize btn-danger"
        type="submit"
      >
        delete
      </button>
    </div>
  </form>
</div>
`;

exports[`CommodityForm renders as expected with update 1`] = `
<div>
  <form>
    <fieldset
      class="my-5 col-start-8 col-span-2"
    >
      <label
        class="inline-block text-sm mb-2"
        for="namespaceInput"
      >
        Type
      </label>
      <div
        class="selector css-b62m3t-container"
        id="namespaceInput"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-3-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
          role="log"
        />
        <div
          class="selector__control css-13cymwt-control"
        >
          <svg
            class="absolute text-md left-[10px]"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 18a7.952 7.952 0 0 0 4.897-1.688l4.396 4.396 1.414-1.414-4.396-4.396A7.952 7.952 0 0 0 18 10c0-4.411-3.589-8-8-8s-8 3.589-8 8 3.589 8 8 8zm0-14c3.309 0 6 2.691 6 6s-2.691 6-6 6-6-2.691-6-6 2.691-6 6-6z"
            />
          </svg>
          <div
            class="selector__value-container css-hlgwow"
          >
            <div
              class="selector__placeholder css-1jqq78o-placeholder"
              id="react-select-3-placeholder"
            >
              &lt;commodity type&gt;
            </div>
            <div
              class="!text-inherit selector__input-container css-19bb58m"
              data-value=""
            >
              <input
                aria-activedescendant=""
                aria-autocomplete="list"
                aria-describedby="react-select-3-placeholder"
                aria-expanded="false"
                aria-haspopup="true"
                aria-label="namespaceInput"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class="selector__input"
                id="react-select-3-input"
                role="combobox"
                spellcheck="false"
                style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="selector__indicators css-1wy0on6"
          >
            <span
              class="hidden selector__indicator-separator css-1u9des2-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class="selector__indicator selector__dropdown-indicator css-1xc3v61-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
        <input
          name="namespaceInput"
          type="hidden"
          value=""
        />
      </div>
      <p
        class="invalid-feedback"
      />
    </fieldset>
    <div
      class="grid grid-cols-12 text-sm gap-2"
    >
      <fieldset
        class="col-span-6"
      >
        <label
          class="inline-block text-sm my-2"
          for="mnemonicInput"
        >
          Code
        </label>
        <input
          class="w-full m-0"
          id="mnemonicInput"
          name="mnemonic"
          type="text"
        />
        <p
          class="invalid-feedback"
        />
      </fieldset>
      <fieldset
        class="col-span-6"
      >
        <label
          class="inline-block text-sm my-2"
          for="fullNameInput"
        >
          Name
        </label>
        <input
          class="w-full m-0"
          id="fullNameInput"
          name="fullname"
          type="text"
        />
        <p
          class="invalid-feedback"
        />
      </fieldset>
    </div>
    <div
      class="flex w-full justify-center mt-5"
    >
      <button
        class="btn capitalize btn-warn"
        type="submit"
      >
        update
      </button>
    </div>
  </form>
</div>
`;
